xkbcat.c

```diff
@@ -142,9 +142,17 @@ int main(int argc, char * argv[]) {
                     char *str = XKeysymToString(s);
                     if (NULL == str) continue;
 
+                    if (cookie->evtype == XI_RawKeyRelease && ev->deviceid != ev->sourceid) continue;
+
                     // Output line
-                    if (printKeyUps) printf("%s",
-                            cookie->evtype == XI_RawKeyPress ? "+" : "-");
+                    if (printKeyUps) {
+                        if (cookie->evtype == XI_KeyPress) {
+                            XIDeviceEvent* dev_ev = cookie->data;
+                            printf("%s", (dev_ev->flags & XIKeyRepeat) ? "=" : "+");
```

Comment:
According to output of `xinput --test-xi2 --root` , `XIDeviceEvent` was the correct type to handle `XI_KeyPress`